numpy/ma/core.py

```diff
@@ -3190,20 +3195,31 @@ def astype(self, newtype):
          [7 -- 9]]
 
         """
-        newtype = np.dtype(newtype)
-        newmasktype = make_mask_descr(newtype)
+        output = super(MaskedArray, self).astype(dtype,
+            order=order, casting=casting, subok=subok, copy=copy)
+
+        # if no copy was made, there's nothing to do
+        if output is self:
+            return self
+
+        # asked to return a base array, so nothing to decorate
+        if not subok:
+            return output
 
-        output = self._data.astype(newtype).view(type(self))
         output._update_from(self)
 
         if self._mask is nomask:
             output._mask = nomask
         else:
-            output._mask = self._mask.astype(newmasktype)
+            # we know we copied the data here, so copy the mask too
+            output._mask = self._mask.astype(
+                make_mask_descr(output.dtype),
+                order=order, casting=casting, subok=subok, copy=True
```

Comment:
Perhaps `assert subok and copy` would be good here then, and then passing constants as the arguments